Issue #4321 Refactored Graceful shutdown

gregw opened the pull request:
removed stopTimeout from all abstractLifeCycles.  It is on Graceful.LifeCycle, which is only implemented by components that can start a graceful shutdown (eg Server, ContextHandler and QueuedThreadPool)

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

removed stopTimeout from all abstractLifeCycles.  It is on Graceful.LifeCycle, which is only implemented by components that can start a graceful shutdown (eg Server, ContextHandler and QueuedThreadPool)

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lachlan-roberts commented:
@gregw the commit could not be built on CI.

> Compilation failure
> 
> [ERROR] /home/jenkins/agent/workspace/tty-10.0.x-4321-GracefulRefactor/jetty-servlet/src/test/java/org/eclipse/jetty/servlet/ServletContextHandlerTest.java:[122,19] org.eclipse.jetty.servlet.ServletContextHandlerTest.MySCIStarter is not abstract and does not override abstract method getStopTimeout() in org.eclipse.jetty.util.component.Graceful.LifeCycle
>

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/AbstractConnector.java:
```diff
@@ -345,13 +371,27 @@ protected void interruptAcceptors()
     {
-        return _shutdown.isShutdown();
+        Shutdown shutdown = _shutdown;
+        return shutdown == null || _shutdown.isShutdown();
```
Don't use the `_shutdown` field directly here could throw NPE if the value can change to null.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/Server.java:
```diff
@@ -173,22 +174,26 @@ public static String getVersion()
-     * @see org.eclipse.jetty.util.component.ContainerLifeCycle#setStopTimeout(long)
      */
-    @Override
     public void setStopTimeout(long stopTimeout)
```
Should have the `@Override` annotation.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/Server.java:
```diff
@@ -173,22 +174,26 @@ public static String getVersion()
+    }
+
+    @ManagedAttribute("Time in ms to gracefully shutdown the server")
+    public long getStopTimeout()
```
`@Override` annotation.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/AbstractHandlerContainer.java:
```diff
@@ -139,50 +141,51 @@ public void setServer(Server server)
-     * returns A MultiException to which any failures are added or null
-     */
-    protected void doShutdown(List<Future<Void>> futures) throws MultiException
+    public interface GracefulContainer extends Graceful.LifeCycle, Container
```
`AbstractHandlerContainer` seems like a strange place to put the `GracefulContainer` interface and the `shutdown` static method. Why is this not done the way you did  `Graceful.LifeCycle` and have `Graceful.Container`.

gregw replied:
Because `org.eclipse.util.container` does not know about Handlers... but yes it is a kind of strange place... I might make it top level.

gregw replied:
hmmm it no longer needs to know about Handlers!

lachlan-roberts commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -675,10 +689,17 @@ private boolean addCounts(int deltaThreads, int deltaIdle)
+                long update = AtomicBiInteger.encode(threads + deltaThreads, idle + deltaIdle);
+                if (_counts.compareAndSet(encoded, update))
+                    return true;
+            }
```
This change is a little concerning, is this somehow related to the graceful shutdown refactor or just a cleanup?

gregw replied:
It is a minor change that allowed better debugging of graceful shutdown.  It allows the idle count to be maintained while the QTP is stopping

gregw committed:
Issue #4321 Refactored Graceful shutdown

cleanup after review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

reinstate other stop tests (more work to do).

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

Fixes for stop test by improving LocalConnector shutdown handling

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

Removed broken test on LocalConnector that is already tested in GracefulStopTest

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

Fixed all stop tests

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lachlan-roberts commented:
@gregw still not building, looks like there are checkstyle violations.

gregw committed:
Issue #4321 Refactored Graceful shutdown

fixed checkstyle

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-4321-GracefulRefactor

gregw committed:
Issue #4321 Refactored Graceful shutdown

No stopTimeout JMX attribute

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

Dump stopTimeout
test with default stopTimeout

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4321 Refactored Graceful shutdown

USe sendError for 503

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet have a look at this now.  Ping me if you want to hangout about it together.

gregw committed:
Issue #4321 Refactored Graceful shutdown

minor cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-4321-GracefulRefactor

gregw committed:
Issue #4321 Refactored Graceful shutdown

Simplifications after review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

janbartel commented on jetty-jmx/src/test/java/org/eclipse/jetty/jmx/ObjectMBeanTest.java:
```diff
@@ -123,7 +123,7 @@ public void testDerivedAttributes() throws Exception
         assertEquals("com.acme.Derived", derivedInfo.getClassName(), "name does not match");
         assertEquals("Test the mbean stuff", derivedInfo.getDescription(), "description does not match");
-        assertEquals(6, derivedInfo.getAttributes().length, "attribute count does not match");
+        assertEquals(5, derivedInfo.getAttributes().length, "attribute count does not match");
```
I don't understand why this value changed here? Doesn't seem anything to do with shutdown?

gregw replied:
the `stopTimeout` used to be an attribute on every AbstractLifeCycle. It no longer is, hence one less attribute.

janbartel commented on jetty-maven-plugin/src/it/jetty-start-gwt-it/beer-client/src/main/java/org/olamy/App.java:
```diff
@@ -44,9 +44,9 @@
-        + "connection and try again.";
+    private static final String SERVER_ERROR = "An error occurred while " +
+        "attempting to contact the server. Please check your network " +
+        "connection and try again.";
```
Again, odd that this has nothing to do with shutdown. Perhaps you are missing a merge from jetty-10.0.x?

gregw replied:
I started getting checkstyle failures, so I had to fix this.  I think it was a problem with the merge of the change to apply checkstyle to tests?

janbartel commented on jetty-util/src/main/java/org/eclipse/jetty/util/component/Graceful.java:
```diff
@@ -43,47 +46,96 @@
  */
 public interface Graceful
```
The comment at line 40 should be for a CompletableFuture, not just a Future.

janbartel reviewed (approved):
Generally LGTM.

gregw committed:
Issue #4321 Refactored Graceful shutdown

after review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lachlan-roberts please re-review

lachlan-roberts commented on jetty-util/src/main/java/org/eclipse/jetty/util/component/Graceful.java:
```diff
@@ -43,47 +47,96 @@
+     * Shutdown the component. When this method returns, the component should not accept any new load.
+     * @return A future that is completed once all load on the component is completed
+     */
+    CompletableFuture<Void> shutdown();
```
Does this really need to be a `CompleteableFuture` rather than just a `Future`?
Would you ever expect the caller of this to complete the future or is it to be done internally like in the `Shutdown` utility with `check()`.

gregw replied:
I guess the whole point of the API is that we don't know how the caller will wait for the shutdown to be complete... and it is possible to use a CompetableFuture as a Future, but not the other way around.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/Server.java:
```diff
@@ -466,21 +464,20 @@ protected void doStop() throws Exception
             {
-                futures.add(connector.shutdown());
+                long end = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(getStopTimeout());
+                Graceful.shutdown(this).get(getStopTimeout(), TimeUnit.MILLISECONDS);
```
I think this should be the only line in the try/catch, otherwise if it throws the stopTimeout will never be set on the ThreadPool.

gregw replied:
ack

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/Server.java:
```diff
@@ -466,21 +464,20 @@ protected void doStop() throws Exception
+                Graceful.shutdown(this).get(getStopTimeout(), TimeUnit.MILLISECONDS);
+                QueuedThreadPool qtp = getBean(QueuedThreadPool.class);
+                if (qtp != null)
+                    qtp.setStopTimeout(Math.min(1000L, TimeUnit.NANOSECONDS.toMillis(end - System.nanoTime())));
```
This logic seems wrong to me, we can never set the stopTimeout on the ThreadPool to longer than 1000ms, and if the `Graceful.shutdown` takes a full stopTimeout we could get a negative value for the ThreadPool stopTimeout.

gregw replied:
oops

lachlan-roberts reviewed (changes requested):
Looks good for the most part, I just have a few more comments and questions first.

gregw committed:
Issue #4321 Refactored Graceful shutdown

after review

Signed-off-by: Greg Wilkins <gregw@webtide.com>
